ballista/core/src/replicator/mod.rs

```diff
@@ -0,0 +1,289 @@
+use object_store::{path::Path, ObjectStore};
+use tokio::io::AsyncWriteExt;
+use tokio::{fs::File, sync::mpsc};
+use tracing::{info, warn};
+
+use crate::error::BallistaError;
+use tokio_util::compat::{Compat, TokioAsyncReadCompatExt};
+
+use self::async_reader::AsyncFileReader;
+pub mod async_reader;
+
+pub enum Command {
+    Replicate { path: String },
+}
+
+pub async fn start_replication(
+    base_path: String,
+    object_store: Arc<dyn ObjectStore>,
+    mut receiver: mpsc::Receiver<Command>,
+) -> Result<(), BallistaError> {
+    while let Some(Command::Replicate { path }) = receiver.recv().await {
+        let destination = format!("{}{}", base_path, path);
+        info!(destination, "Start replication");
+
+        match Path::parse(destination) {
+            Ok(dest) => match load_file(path.as_str()).await {
+                Ok(reader) => {
+                    if let Err(error) =
+                        replicate_to_object_store(&dest, reader, object_store.clone())
+                            .await
+                    {
+                        warn!(?error, ?path, "Failed to upload file to object store");
+                    }
+                }
+                Err(error) => {
+                    warn!(?error, ?path, "Failed to open file");
+                }
+            },
+            Err(error) => {
+                warn!(?error, ?path, "Failed to parse replication path");
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn load_file(path: &str) -> Result<AsyncFileReader<Compat<File>>, BallistaError> {
+    let file = File::open(path).await?;
+    let reader = AsyncFileReader::try_new(file.compat(), None).await?;
+
+    Ok(reader)
+}
+
+fn serialize_batch(batch: RecordBatch) -> Result<bytes::Bytes, BallistaError> {
+    let mut serialized_data = Vec::with_capacity(batch.get_array_memory_size());
+
+    let options = IpcWriteOptions::default()
+        .try_with_compression(Some(CompressionType::LZ4_FRAME))
+        .unwrap_or(IpcWriteOptions::default());
+
+    {
+        let mut writer = StreamWriter::try_new_with_options(
+            &mut serialized_data,
+            batch.schema().as_ref(),
+            options,
+        )?;
+
+        writer.write(&batch)?;
+        writer.finish()?;
+    }
+
+    Ok(Bytes::from(serialized_data))
+}
+
+async fn replicate_to_object_store(
+    destination: &Path,
+    mut reader: AsyncFileReader<Compat<File>>,
+    object_store: Arc<dyn ObjectStore>,
+) -> Result<(), BallistaError> {
+    let (_, mut upload) = object_store.put_multipart(destination).await.map_err(|e| {
+        BallistaError::General(format!("Failed to create object store writer: {:?}", e))
+    })?;
+
+    while let Some(batch) = reader.maybe_next().await.transpose() {
+        if let Ok(batch) = batch {
+            let data = serialize_batch(batch)?;
+            upload.write_all(&data).await?;
+        }
+    }
+
+    upload.flush().await?;
+    upload.shutdown().await?;
```

Comment:
We might need to tweak this a bit. The minimum part size for S3 is 5MiB (except for the last part). I don't believe that the `object_store` does any buffering internally so writing each batch as a part might not work correctly

Reply:
It does internal buffering